### PR TITLE
TW-15725

### PR DESCRIPTION
**Pull request by dlombardi**
When reviewing this PR any file beginning with `dist/` can be ignored for review purposes.  That's just the transpiled and source mapped code.

Fixes a few things:
1. operations are now placed where they belong in redux instead of a level above where they should have been.
2. every operation now includes its operationKey for convenient access within redux connected modules.
3.  Introduces asyncOperationStep callbacks in descriptor definitions.  These callbacks will be invoked on Begin, Resolve, Reject of all operations associated with the descriptor. Params are passed through the args. Invalidating operations with `invalidateAsyncOperation` can be performed in these callbacks.
4. An API `invalidateAsyncOperation` is now available to invalidate an operation directly by providing a descriptorId and the params

Both `invalidateAsyncOperation` API and the resolve callback using invalidateAsyncOperation are unit-tested

### CODE REVIEW

File: src/asyncOperationManagerUtils.js [hunk 1]
```diff
@@ -61,7 +61,12 @@ const registerAsyncOperationDescriptors = (asyncOperationDescriptors, ...otherDe
   return asyncOperationManagerState.setState(newState);
 };
 
-const getAsyncOperation = (state, descriptorId, params, otherFields) => {
+const getAsyncOperation = ({
+  state,
+  descriptorId,
+  params,
+  otherFields,
+}) => {
```
Comment (AceNorth) [inline src/asyncOperationManagerUtils.js]: Is the plan to have a second PR to update all the usages of `getAsyncOperation` in the project?

Comment (dlombardi) [reply src/asyncOperationManagerUtils.js]: great catch.  I'll change it back to it doesn't need to change throughout the application

Comment (AceNorth) [reply src/asyncOperationManagerUtils.js]: Oh, ok! Are you going to have `getAsyncOperationFromState` take an object still?

File: src/asyncOperationManagerUtils.js [hunk 2]
```diff
@@ -61,7 +61,18 @@ const registerAsyncOperationDescriptors = (asyncOperationDescriptors, ...otherDe
   return asyncOperationManagerState.setState(newState);
 };
 
-const getAsyncOperation = (state, descriptorId, params, otherFields) => {
+const invalidateAsyncOperationByKey = (asyncOperationKey, descriptorId) => {
```
Comment (fawcilize) [inline src/asyncOperationManagerUtils.js]: It seems like we could do `invalidateAsyncOperation = (descriptorId, asyncOperationKey)`

Where asyncOperationKey is optional and targets the specific operation and if left out, invalidates all of the descriptor's operations. That would eliminate the need for `invalidatingOperationsDescriptorIds` and leave the invalidating solely up to the consumer, simplifying this library's role.

File: src/asyncOperationStateUtils.js [hunk 3]
```diff
@@ -52,10 +55,44 @@ const updateAsyncOperationDescriptor = (state, descriptorOptions) => {
...
+      }
+      if (acc.lastFetchStatusTime <= operation.lastFetchStatusTime) {
+        return operation;
+      }
+      return acc;
+    }
+  }, {});
+};
+
+const setInvalidatedOperationByKeyState = (state, asyncOperationKey, descriptorId) => {
```
Comment (fawcilize) [inline src/asyncOperationStateUtils.js]: It's more like `createInvalidatedOperationByKeyState`, or even `createInvalidatedOperationState`

set seems like a confusing word since it's not actually setting the state.

File: src/asyncOperationStateUtils.js [hunk 4]
```diff
@@ -52,10 +55,44 @@ const updateAsyncOperationDescriptor = (state, descriptorOptions) => {
...
+        : initialWriteAsyncOperationForAction(asyncOperationDescriptor.descriptorId, asyncOperationKey),
+    },
+  };
+};
+
 // This function will do all the work to determine if an async operation is returned as an initial async operation
 // (if it is not found in state), an asyncOperation with parentAsyncOperation metaData (recursively searched to find if the parentAsyncOperation is more
 // up-to-date) or just the asyncOperation itself if the none of the above apply.
-const getAsyncOperation = (state, asyncOperationKey, asyncOperationDescriptor, asyncOperationParams, fieldsToAdd) => {
+const getAsyncOperationFromState = ({
```
Comment (fawcilize) [inline src/asyncOperationStateUtils.js:79]: What's the purpose of this change? It seems like it would be unnecessarily breaking.

**Review by fawcilize — CHANGES_REQUESTED** [review]

File: src/asyncOperationManagerUtils.js (hunk 1 above)
Comment (dlombardi) [reply src/asyncOperationManagerUtils.js]: yep!  It seems like it wouldn't hurt

File: src/asyncOperationStateUtils.js (hunk 4 above)
Comment (dlombardi) [reply src/asyncOperationStateUtils.js:79]: it changes the name so it doesn't conflict with `getAsyncOperation` that is exported by asyncOperationManagerUtils which is the actual api exposed to the consumer. 

This shouldn't break consumer implementations

File: src/asyncOperationStateUtils.js (hunk 3 above)
Comment (dlombardi) [reply src/asyncOperationStateUtils.js]: it's now `createInvalidatedOperationState`

File: src/asyncOperationManagerUtils.js (hunk 2 above)
Comment (dlombardi) [reply src/asyncOperationManagerUtils.js]: we could still implement this if it's still desired.

File: src/asyncOperationManagerUtils.js [hunk 5]
```diff
@@ -131,14 +154,32 @@ const getStateForOperationAfterStep = (state, asyncOperationStep, descriptorId,
     otherFields,
   } = getAsyncOperationInfo(newState.descriptors, descriptorId, params);
 
-  // in case operation/descriptor state is initialized in userland we pass that through
-  // to the library state.
+  // descriptor asyncOperationStep callbacks
+  if (asyncOperationDescriptor.onBegin && asyncOperationStep === ASYNC_OPERATION_STEPS.BEGIN_ASYNC_OPERATION) {
```
Comment (fawcilize) [inline src/asyncOperationManagerUtils.js]: Not that it's really important, but this will never be all three steps at the same time, so it could either be an if/else or a switch (probably nicer syntax).

**Review by fawcilize — APPROVED** [review]

**Review by nickpassarella — APPROVED** [review]
